flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.fs;
+
+import org.apache.flink.util.ChildFirstClassLoader;
+import org.apache.flink.util.TemporaryClassLoaderContext;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.net.URL;
+import java.security.AccessControlContext;
+import java.security.AccessController;
+import java.security.ProtectionDomain;
+import java.util.List;
+import java.util.Vector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** Tests for the {@link SafeFileSystemFactory}. */
+public class SafeFileSystemFactoryTest {
```

Comment:
Maybe, it's the time of day, but I struggle to follow what the test is doing. I'm gonna try to go over it once more tomorrow morning. But it might be an indication that we could add a few more comments?

Reply:
See https://github.com/apache/flink/pull/17916#discussion_r761288714 for  explanation.

Once we're sure that this is the approach we want to take, I'll try to come up with more explanatory javadocs.

We should also get more people to validate that we really want to do this before merging. We're still not sure with @zentol that the outcome outweights the complexity of this change.

Reply:
What are your concerns about the approach you implemented in this PR? 🤔

Reply:
Mostly it's about... lot of hustle for a "minor" issue. We'll leak at most one class loader per process without this fix, which should be bearable as it adds a constant overhead to the memory requirements.

But on the other hand, we shouldn't ideally leak anything and the final solution is not as "hacky" as the previous approaches. 🤔

Reply:
I wouldn't consider it "too hacky" if we add some more documentation. 👍